models/classes/resources/Service/ClassDeleter.php

```diff
@@ -0,0 +1,127 @@
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; under version 2
+ * of the License (non-upgradable).
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2021 (original work) Open Assessment Technologies SA;
+ */
+
+declare(strict_types=1);
+
+namespace oat\tao\model\resources\Service;
+
+use core_kernel_classes_Class;
+use core_kernel_classes_Property;
+use core_kernel_classes_Resource;
+use oat\generis\model\data\Ontology;
+use oat\tao\model\search\index\OntologyIndex;
+use oat\tao\model\accessControl\PermissionCheckerInterface;
+
+class ClassDeleter implements ClassDeleterInterface
+{
+    private const PROPERTY_INDEX = OntologyIndex::PROPERTY_INDEX;
+
+    /** @var PermissionCheckerInterface */
+    private $permissionChecker;
+
+    /** @var Ontology */
+    private $ontology;
+
+    public function __construct(PermissionCheckerInterface $permissionChecker, Ontology $ontology)
+    {
+        $this->permissionChecker = $permissionChecker;
+        $this->ontology = $ontology;
+    }
+
+    public function delete(core_kernel_classes_Class $class, core_kernel_classes_Class $rootClass): void
+    {
+        if (!$class->equals($rootClass)) {
+            $this->deleteClassRecursively($class);
+        }
+    }
+
+    public function isDeleted(core_kernel_classes_Class $class): bool
+    {
+        return !$class->exists();
+    }
+
+    private function deleteClassRecursively(core_kernel_classes_Class $class): bool
+    {
+        $isClassDeletable = true;
+
+        foreach ($class->getSubClasses() as $subClass) {
+            $isClassDeletable = $this->deleteClassRecursively($subClass) && $isClassDeletable;
+        }
+
+        return $this->deleteClass($class, $isClassDeletable);
+    }
+
+    private function deleteClass(core_kernel_classes_Class $class, bool $isClassDeletable): bool
+    {
+        return $this->deleteInstances($class->getInstances())
+            && $isClassDeletable
+            && $this->deleteProperties($class->getProperties())
+            && $class->delete();
+    }
+
+    /**
+     * @param core_kernel_classes_Resource[] $instances
+     */
+    private function deleteInstances(array $instances): bool
+    {
+        $status = true;
+
+        foreach ($instances as $instance) {
+            if (!$this->permissionChecker->hasWriteAccess($instance->getUri())) {
+                $status = false;
+
+                continue;
+            }
+
+            $instance->delete();
```

Comment:
nitpick: Maybe using a standard `if`/`else` branching condition here makes it more readable instead jumping with `continue;` ?

```suggestion
            if (!$this->permissionChecker->hasWriteAccess($instance->getUri())) {
                $status = false;
            } else {
                $instance->delete();
            }
```

Reply:
We can even invert the condition and put the "happy path" first:

```suggestion
            if ($this->permissionChecker->hasWriteAccess($instance->getUri())) {
                $instance->delete();
            } else {
                $status = false;
            }
```

Reply:
We do not use else as recommendation of code calisthenics. https://williamdurand.fr/2013/06/03/object-calisthenics/#2-dont-use-the-else-keyword

You never need else.

Reply:
Then I'd use something like your last example (however `$status` should be kept as false if at least one call to `hasWriteAccess()` returns `false`, since it is used as the return value of the method as a whole).

```suggestion
            if ($this->permissionChecker->hasWriteAccess($instance->getUri())) {
                $instance->delete();
                continue;
            }
            
            $status = false;         
```

Reply:
I updated my example. The `deletion` status should also be considered IMO. 

Ping @shpran

Reply:
Ping @hectoras @gabrielfs7 
```php
private function deleteInstances(array $instances): bool
{
    $status = true;

    foreach ($instances as $instance) {
        if (!$this->permissionChecker->hasWriteAccess($instance->getUri()) || !$instance->delete()) {
            $status = false;
        }
    }

    return $status;
}
```